feat(eslint-plugin): add explicit-module-boundary-types rule

GuyPie opened the pull request:
Fixes #65

phaux commented:
All the other options are `allowSomething` and only this one is `ignoreSomething`. Is there a difference? If not, I think it should be like all the others

GuyPie commented:
@phaux makes sense, I renamed it.

bradzacher commented:
Hmm. I wonder if this is better suited being combined with [`no-untyped-public-signature`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-untyped-public-signature.md) as a standalone rule, as opposed to an option on this rule.

I feel like having a single rule that says "ensure the module boundary is explicitly typed" is probably better than a collection of rules with options.

GuyPie commented:
I agree. How about replacing "no-untyped-public-signature" with a new "explicit-module-boundary-types" rule?

bradzacher commented:
that sounds like a decent idea.
You can use the `deprecated`, and `replacedBy` `meta` props to point users at the new rule.

Then it's just a matter of copying the code across. We can delete the old rule in the next major.

You can probably make the return type checks a lot simpler as well. Right now `explicit-function-return-type` is pretty complex because it has a lot of "I'll ignore this case, as a human can probably infer the type here".

To start with, the rule can probably just care about these simple cases to start with:
```ts
export [default] class Foo {
  method(arg) {} // error - need explicit argument type
                 // error - need explicit function return type

  fnMember1 = (arg) => {} // error - need explicit argument type
                          // error - need explicit function return type

  fnMember2 = function foo(arg) {} // error - need explicit argument type
                                   // error - need explicit function return type

  get getter() {} // error - need explicit function return type
  set setter(arg) {} // error - need explicit argument type
}

export default {
  // same as above, but an object
}
export const obj = {
  // ditto
}

export [default] function bar1(arg) {} // error - need explicit argument type
                                       // error - need explicit function return type
export const bar2 = function (arg) {} // error - need explicit argument type
                                      // error - need explicit function return type

export const baz = (arg) => {} // error - need explicit argument type
                               // error - need explicit function return type
export default (arg) => {} // error - need explicit argument type
                           // error - need explicit function return type
```

is there anything I missed there...?

GuyPie commented:
@bradzacher I added the new `explicit-module-boundary-types` rule, let me know what you think. In the end it's pretty much a direct copy of `explicit-function-return-type`, with a few changes - 

* It only checks exported functions and methods of exported classes.
* I removed the `allowExpressions` option, since it's not applicable to module boundaries.
* I added the `allowedNames` option, which is a string array of function/method names to skip the check for.

Let me know what you think.

GuyPie commented on packages/eslint-plugin/tests/configs/all.test.ts:
```diff
@@ -9,15 +9,10 @@ describe('all.json config', () => {
+  const prefixedRuleNames = rulesNames.reduce<string[]>((collection, name) => {
+    collection.push(`${RULE_NAME_PREFIX}${name}`);
+    return collection;
+  }, []);
```
It broke once I deprecated a rule, I think the test is wrong since it appears deprecated rules are in face being included.

bradzacher replied:
They shouldn't be - https://github.com/typescript-eslint/typescript-eslint/blob/93390e6100144740cd1cac76960234bc4239bcdb/packages/eslint-plugin/tools/generate-configs.ts#L144-L154

Could you please try running `yarn generate:configs` again?

GuyPie replied:
That fixed it, I was not aware I need to run it. Thanks!

GuyPie commented:
> one thing that's missing is the argument handling from the `no-untyped-public-signature` rule.
> 
> Eg:
> 
> ```ts
> export default const (arg): void => {} // error - need explicit argument type
> 
> export function foo(arg): number { return 1; } // error - need explicit argument type
> 
> export class Foo {
>   bar(arg): string { return '' } // error - need explicit argument type
> }
> ```
> 
> could you please copy that over as well?

👍 will do.

GuyPie commented:
@bradzacher I added a 'missingArgType' check=.
